File: frontend/src/types/tutor.ts
/**
 * TypeScript type definitions for AI Tutor feature.
 */

export type LearningMode = "free_talk" | "quiz" | "weak_point";

export type SessionStatus = "active" | "ended" | "timed_out";

export interface TutorMessage {
  role: "user" | "assistant";
  content: string;
  related_cards: string[];
  timestamp: string;
}

export interface TutorSession {
  session_id: string;
  deck_id: string;
  mode: LearningMode;
  status: SessionStatus;
  messages: TutorMessage[];
  message_count: number;
  created_at: string;
  updated_at: string;
  ended_at: string | null;
}

export interface StartSessionRequest {
  deck_id: string;
  mode: LearningMode;
}

export interface SendMessageRequest {
  content: string;
}

export interface SendMessageResponse {
  message: TutorMessage;
  session_id: string;
  message_count: number;
  is_limit_reached: boolean;
}

export interface SessionListResponse {
  sessions: TutorSession[];
}

File: backend/src/models/tutor.py
"""Pydantic models for AI Tutor feature."""

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field


class TutorMessage(BaseModel):
    """Individual message within a tutor session."""

    role: Literal["user", "assistant"]
    content: str
    related_cards: list[str] = Field(default_factory=list)
    timestamp: datetime


class TutorSessionResponse(BaseModel):
    """Session data returned to the client."""

    session_id: str
    deck_id: str
    mode: Literal["free_talk", "quiz", "weak_point"]
    status: Literal["active", "ended", "timed_out"]
    messages: list[TutorMessage]
    message_count: int
    created_at: datetime
    updated_at: datetime
    ended_at: Optional[datetime] = None


class StartSessionRequest(BaseModel):
    """Request to start a new tutor session."""

    deck_id: str = Field(..., min_length=1)
    mode: Literal["free_talk", "quiz", "weak_point"]


class SendMessageRequest(BaseModel):
    """Request to send a message in a tutor session."""

    content: str = Field(..., min_length=1, max_length=2000)


class SendMessageResponse(BaseModel):
    """Response after sending a message."""

    message: TutorMessage
    session_id: str
    message_count: int
    is_limit_reached: bool


class SessionListResponse(BaseModel):
    """List of tutor sessions."""

    sessions: list[TutorSessionResponse]
    total: int


LearningMode = Literal["free_talk", "quiz", "weak_point"]
